### DISCUSSION TITLE
How to map a .Xmodmap config to a keybind?

### DISCUSSION THREAD

**Discussion by hamensman** [qtile/qtile#2792] [Q&A] [answered]
How would you create a keybind which saves the changes of  `.Xmodmap` onto the system? Normally, you create a `.Xmodmap`, write whatever keymap modifications you want, then save the changes with `xmodmap ~/.Xmodmap` on the terminal.

If I write, say, `Key([], "Alt_L", lazy.spawn("xmodmap ~/.Xmodmap")),` on my qtile config, this won't do anything. What do I need to write instead?

**Comment by Fabien-Bernard** [comment, marked as answer]
I think the `~` is not interpreted as in your terminal. 
I read some config examples using `os.path.expanduser("~")`

**Reply by hamensman** [reply]
You're awesome! I had to write the longhand version of it (`/home/user/.Xmodmap`) and it worked! Thanks a lot!

I thought of something else I'm curious about.

How can I (I suppose write a function that allows me to) toggle between, say `.Xmodmap` and `.Xmodmap2` and `.Xmodmap3` using the same key? Basically how a lot of mechanical keyboards have programmable keyboards "layers" which can be customised on their software. I wonder how that can be done purely on the qtile side, or if the qtile team would consider making such a widget (not sure if this can already be done with the `KeyboardLayout` widget...)

**Reply by Fabien-Bernard** [reply]
I think the KeyboardLayout widget is only using predefined layouts, so I guess you'll have to create your owns before being able to cycle through them with the widget.

To do it simply with a function in your config, you could create an iterable containing your different xmodmaps, then a function that cycle through it and launch xmodmap. Finally bind the function to your key.

Something like : 
```
from itertools import cycle
from os.path import expanduser
from libqtile import qtile

home = expanduser("~") + "/"
my_xmodmaps = [".xmodmap1", ".xmodmap2", ".xmodmap3"]
xmm_cycle = cycle(my_xmodmaps)
                          
def toggle_xmodmap(qtile):                                                                 
    current_xmodmap = home + next(xmm_cycle)                                          
    qtile.cmd_spawn("xmodmap " + current_xmodmap)

Key([], "Alt_L", lazy.function(toggle_xmodmap))
```

**Reply by hamensman** [reply]
Thanks a lot. Highly appreciate the help!
